lib/site_prism/dsl.rb

```diff
@@ -205,13 +205,19 @@ def create_helper_method(proposed_method_name, *find_args)
         end
       end
 
-      def create_rspec_existence_matchers(element_name)
+      def create_rspec_existence_matchers(element_name) # rubocop:disable Metrics/MethodLength
         matcher = "has_#{element_name}?"
         negated_matcher = "has_no_#{element_name}?"
 
         RSpec::Matchers.define "have_#{element_name}" do |*args|
           match { |actual| actual.public_send(matcher, *args) }
-          match_when_negated { |actual| actual.public_send(negated_matcher, *args) }
+          match_when_negated do |actual|
+            if actual.respond_to?(negated_matcher)
+              actual.public_send(negated_matcher, *args)
```

Comment:
Looking at your spec below, this seems a proper edge case, and I'd like to learn more about a realistic use case.

I've nothing against the logic, more about "why" someone would code a conflicting boolean with the method invocation

Reply:
Ok, so more context on how this actually came up:

We have a `Customer` model, and that model has a `#has_notification?` method defined on it. We have and existing test that asserts

```ruby
  expect(customer).not_to have_notification(:something_happened)
```

rspec provides these ~existence~ predicate matchers (I had the terminology wrong) out of the box: https://relishapp.com/rspec/rspec-expectations/v/2-4/docs/built-in-matchers/predicate-matchers

Many moons later, I was writing a test using site prism for something completely separate, a page which displays notifications.

```ruby
class MyPage < SitePrism::Page
  element :notification, :xpath, '//a[@class="notifications"]'
end
```

The test I wrote worked perfectly, but when I ran the whole test suite, the existing tests that used the standard rspec negated predicate matchers were broken in surprising ways

```
1) Customer is expected not to have notification :something_happened

 Failure/Error: expect(customer.reload).to_not have_notification(:something_happened)

 NoMethodError:
 undefined method `has_no_notification?' for #<Customer:0x0000559dc92bfdd8>
```

It took quite a bit of splunking to get to the bottom of how/why adding an element on a site prism object could even affect this test, but I'm pretty confident that replacing the standard matcher [here](https://github.com/site-prism/site_prism/pull/50/files#diff-909e3d337607154302f0efd4be953dc8L214) is the root cause.

Reply:
Can you address the comment below (I think I'm wrong, as the localisation aspect of your caller is different - One using Customer Page, the other using AnotherPage)

Also given this should be quite rare, and this is a bit muddy in terms of what "type" of fix this is. I think we definitely should add some logging. `SitePrism.logger.LEVEL` is what you want to use here.

Reply:
Sorry @luke-hill, I'm not sure I understand re: the localisation aspect. Rspec matchers are unscoped as far as I can tell, and the ones site_prism is defining break some of the ones that are part of standard rspec

I put together this sample repo to try and better illustrate since I don't think I'm explaining myself very clearly

https://github.com/lparry/rspec-site-prism-bug-demo/pull/4

If you take a look, I have a repo with out-of-the-box rspec and an object that I'm testing using out of the box rspec helpers [here](https://github.com/lparry/rspec-site-prism-bug-demo/pull/4/commits/efc069040852942cd1baa1e5f586e004ad093920) which is green

In the [next commit](https://github.com/lparry/rspec-site-prism-bug-demo/pull/4/commits/0fb37c7475477a96ebfc4b4fb0fe76882a2c753e), I add site prism and define a site prism page class (hypothetically to test some completely unrelated page, but I'm only defining the class here, not even using it), and defining this class causes the out of the box rspec matchers to be overridden with ones that are incompatible with my previously green test, and causes a red build.

I'm sorry if I'm being dense, but I'm unclear what kind of logging you're getting at. Not breaking unrelated rspec tests seems like a pretty straightforward bugfix, I don't follow when we would need to log or what we would output

Reply:
So I'll copy it in here.

SitePrism defines these boolean checks on the fly. What I'm saying is your current approach will "already" have a race condition. It feels like you want one behaviour for one race condition and another behaviour for another?

So essentially, inside SitePrism we define these custom rspec matchers. Other people have noticed odd behaviour with them as well (Because they are procedural), and often inside a stack, the top requires are done in an arbitrary order. What I'm saying is, if you have a mix of SitePrism pages and non siteprism pages, and you think this is causing you an issue. Shouldn't we address the problem "head-on" by simply configuring where and when these methods are created?

As for your repo / demonstration, I fully understand "how" the issue is occurring, that's not the issue. Just by fixing it one way it potentially alters behaviour for others. I'd rather understand that since you're encountering the issue one way, lets put some logging in to help diagnose this issue for others when it crops up (Which I envisage it will), down the line

Reply:
Sorry Luke, I sense youre getting frustrated with me but I assure you I'm not trying to be patronising, I'm honestly just not following.

I don't think I understand the race condition. In my mental model things work like this:

* bundler loads up rspec and site prism and the rest of the gemfile
* rspec comes with some magic to metaprogram `has_<foo>?` matchers into its collection of matchers
* defining a site prism class overrides those out-of-the-box matchers with a version that expects objects to respond to both `#has_foo?` and `#has_no_foo?`, but `#has_no_foo?` is not a usual convention in Rails apps.
* unrelated unit tests that were previously passed start failing

I've tried reordering the Gemfile and reorder the definitions in my example repo [here](https://github.com/lparry/rspec-site-prism-bug-demo/blob/0f41011b3f835d5528b69bc963e5b004fe4fffce/spec/example_spec.rb) but they continue to fail in the same manner. Can you help me to better understand where my mental model is mismatched, or how/when the "race condition" might result in the opposite problem?

The "breaking change"/no race condition solution feels like "dont reuse matcher prefixes that already exist in standard rspec", but I can see how that just pushes the problem to every user. It seems like not calling methods on an object that don't exist is the next best thing, although if it is possible for these to fail for the opposite reason (rspec "overriding" site prism matchers?) I can see how this fix would only solve half the problem.

Alternatively line of thought: are we doing something wrong by mixing our site prism tests with the rest of our test suite? You've referred to non-siteprism pages a few time, but that's not what's actually being affected. The affected tests are plain old unit tests. Should site prism be isolated from the rest of our test suite instead of all being run using the same `spec_helper.rb` so that these matchers never have a chance to interact with the ones expected in our unit tests?

Reply:
Oki so I think you've spotted the race condition here. Because what's happening is you mention `rspec` comes in step 2 then site prism comes in step 3. Those steps are controlled by yourself, they could happen in any order. Having conflicts like this is part and parcel of ruby.

I'm not against fixing it, but just want you to understand that first.

The fix you've suggested is fine. But it needs documenting (In terms of log messages), that we're encountering a conflict. Because we are essentially conditionally doing metaprogramming. At runtime, that's something that's going to be super hard to triage.

At the moment, we know exactly what the issue is, in 3 months time, I will have forgotten about it.

The re-ordering will lead to some of the methods working in different ways, but as you always have a defined `has_foo?` you won't hit an NME. Lets just add some tidy ups here and add logging, and we can always revisit later on if other people encounter the same issue.